File: sdk/python/tempmail_sdk/providers/fakemail.py
"""
FakeMail 渠道 -- https://www.fakemail.net
"""

import json
import re
from typing import Any, Dict, List, Optional

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "fakemail"
BASE_URL = "https://www.fakemail.net"
CSRF_RE = re.compile(r'const\s+CSRF\s*=\s*"([^"]+)"')


def _merge_cookie(prev: str, resp: Any) -> str:
    jar: Dict[str, str] = {}
    for part in (prev or "").split(";"):
        p = part.strip()
        if "=" in p:
            k, v = p.split("=", 1)
            jar[k] = v
    for cookie in resp.cookies:
        jar[cookie.name] = cookie.value
    return "; ".join(f"{k}={v}" for k, v in jar.items())


def _json(text: str) -> Any:
    return json.loads((text or "").lstrip("\ufeff"))


def _home_headers() -> Dict[str, str]:
    return {"Accept": "text/html,application/xhtml+xml", "User-Agent": "Mozilla/5.0"}


def _ajax_headers(cookie: str) -> Dict[str, str]:
    return {
        "Accept": "application/json, text/javascript, */*; q=0.01",
        "X-Requested-With": "XMLHttpRequest",
        "Referer": f"{BASE_URL}/",
        "Cookie": cookie,
        "User-Agent": "Mozilla/5.0",
    }


def _open_session() -> Dict[str, str]:
    resp = tm_http.get(f"{BASE_URL}/", headers=_home_headers(), timeout=15)
    resp.raise_for_status()
    m = CSRF_RE.search(resp.text)
    if not m:
        raise ValueError("fakemail: csrf token not found")
    return {"csrf": m.group(1), "cookie": _merge_cookie("", resp)}


def _create_address(session: Dict[str, str]) -> Dict[str, Any]:
    resp = tm_http.get(
        f"{BASE_URL}/index/index",
        params={"csrf_token": session["csrf"]},
        headers=_ajax_headers(session["cookie"]),
        timeout=15,
    )
    resp.raise_for_status()
    session["cookie"] = _merge_cookie(session["cookie"], resp)
    data = _json(resp.text)
    if not isinstance(data, dict):
        raise ValueError("fakemail: invalid mailbox response")
    return data


def _fetch_rows(cookie: str) -> List[Dict[str, Any]]:
    resp = tm_http.get(
        f"{BASE_URL}/index/refresh", headers=_ajax_headers(cookie), timeout=15
    )
    resp.raise_for_status()
    data = _json(resp.text)
    return data if isinstance(data, list) else []


def _fetch_detail(cookie: str, message_id: str) -> Optional[Dict[str, Any]]:
    try:
        resp = tm_http.post(
            f"{BASE_URL}/index/email",
            data={"id": message_id},
            headers=_ajax_headers(cookie),
            timeout=15,
        )
        if not resp.ok:
            return None
        data = _json(resp.text)
        return data if isinstance(data, dict) else None
    except Exception:
        return None


def _flatten(
    row: Dict[str, Any], detail: Optional[Dict[str, Any]], recipient: str
) -> Dict[str, Any]:
    detail = detail or {}
    return {
        "id": detail.get("id") or row.get("id"),
        "from": detail.get("od") or row.get("od") or "",
        "to": recipient,
        "subject": detail.get("predmet") or row.get("predmet") or "",
        "text": "",
        "html": detail.get("telo") or "",
        "date": row.get("kdy") or "",
        "isRead": row.get("precteno") == "precteno",
    }


def generate_email() -> EmailInfo:
    session = _open_session()
    data = _create_address(session)
    email = str(data.get("email") or "").strip()
    if not email or "@" not in email:
        raise ValueError("fakemail: invalid mailbox response")
    return EmailInfo(channel=CHANNEL, email=email, _token=session["cookie"])


def get_emails(token: str, email: str) -> List[Email]:
    cookie = (token or "").strip()
    address = (email or "").strip()
    if not cookie:
        raise ValueError("fakemail: empty session token")
    if not address:
        raise ValueError("fakemail: empty email")
    emails: List[Email] = []
    for row in _fetch_rows(cookie):
        if not isinstance(row, dict):
            continue
        message_id = str(row.get("id") or "").strip()
        detail = _fetch_detail(cookie, message_id) if message_id else None
        emails.append(normalize_email(_flatten(row, detail, address), address))
    return emails

File: sdk/ruby/lib/tempmail_sdk/providers/fakemail.rb
# frozen_string_literal: true

require "json"
require "uri"

module TempmailSdk
  module Providers
    # fakemail.net 渠道实现（与 disposablemail.com 共享 PHP API 结构）
    #
    # 流程：
    #   1. GET / 获取 session cookie + HTML 中的 CSRF token（const CSRF = "xxx"）
    #   2. GET /index/index?csrf_token={csrf} 创建邮箱，返回 {"email":"...","heslo":"..."}
    #   3. GET /index/refresh 携带 session cookie 获取邮件列表（捷克语字段: predmet/od/kdy/precteno）
    #   4. POST /index/email（body: id={id}）获取邮件 HTML 正文
    # token 存储 session cookie 字符串
    module Fakemail
      CHANNEL = "fakemail"
      BASE_URL = "https://www.fakemail.net"

      CSRF_RE = /const\s+CSRF\s*=\s*"([^"]+)"/.freeze

      BROWSER_HEADERS = {
        "Accept" => "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
        "User-Agent" => "Mozilla/5.0"
      }.freeze

      AJAX_HEADERS = {
        "Accept" => "application/json, text/javascript, */*; q=0.01",
        "Referer" => "#{BASE_URL}/",
        "User-Agent" => "Mozilla/5.0",
        "X-Requested-With" => "XMLHttpRequest"
      }.freeze

      module_function

      # 合并 Set-Cookie 为 Cookie 字符串
      def merge_cookies(existing, resp)
        jar = {}
        existing.to_s.split(";").each do |part|
          part = part.strip
          next unless part.include?("=")

          k, v = part.split("=", 2)
          jar[k.strip] = v.to_s
        end
        resp.set_cookies.each do |line|
          first = line.split(";", 2).first.to_s.strip
          next unless first.include?("=")

          k, v = first.split("=", 2)
          jar[k.strip] = v.to_s
        end
        jar.map { |k, v| "#{k}=#{v}" }.join("; ")
      end

      # 去除 UTF-8 BOM
      def clean_json(body)
        body = body.dup.force_encoding("BINARY")
        body = body[3..] if body.start_with?("\xEF\xBB\xBF".b)
        body.force_encoding("UTF-8")
      end

      # 创建临时邮箱
      # @return [EmailInfo]
      def generate_email
        r1 = Http.get("#{BASE_URL}/", headers: BROWSER_HEADERS)
        r1.raise_for_status
        cookie = merge_cookies("", r1)
        m = r1.body.match(CSRF_RE)
        raise "fakemail: csrf token not found" unless m

        csrf = m[1]
        r2 = Http.get("#{BASE_URL}/index/index?csrf_token=#{URI.encode_www_form_component(csrf)}",
                      headers: AJAX_HEADERS.merge(cookie.empty? ? {} : { "Cookie" => cookie }))
        r2.raise_for_status
        cookie = merge_cookies(cookie, r2)

        data = JSON.parse(clean_json(r2.body))
        email = data["email"].to_s.strip
        raise "fakemail: invalid mailbox response" if email.empty? || !email.include?("@")

        EmailInfo.new(channel: CHANNEL, email: email, token: cookie)
      end

      # 获取邮件列表
      # @param token [String] session cookie 字符串
      # @param email [String]
      # @return [Array<Email>]
      def get_emails(token, email)
        raise "fakemail: empty session token" if token.to_s.strip.empty?
        raise "fakemail: empty email" if email.to_s.strip.empty?

        headers = AJAX_HEADERS.merge("Cookie" => token)
        resp = Http.get("#{BASE_URL}/index/refresh", headers: headers)
        resp.raise_for_status

        rows = begin
          JSON.parse(clean_json(resp.body))
        rescue JSON::ParserError
          []
        end
        return [] unless rows.is_a?(Array) && !rows.empty?

        rows.map do |row|
          id = row["id"].to_s
          detail = fetch_detail(token, id) unless id.empty?
          subject = detail&.dig("predmet").to_s
          subject = row["predmet"].to_s if subject.empty?
          from_addr = detail&.dig("od").to_s
          from_addr = row["od"].to_s if from_addr.empty?
          html_body = detail&.dig("telo").to_s
          Normalize.normalize_email({
            "id" => id,
            "from" => from_addr,
            "to" => email,
            "subject" => subject,
            "html" => html_body,
            "date" => row["kdy"].to_s,
            "isRead" => row["precteno"] == "precteno"
          }, email)
        end
      end

      # 获取单封邮件详情
      def fetch_detail(cookie, id)
        return nil if id.empty?

        form = "id=#{URI.encode_www_form_component(id)}"
        resp = Http.post("#{BASE_URL}/index/email",
                         headers: AJAX_HEADERS.merge(
                           "Content-Type" => "application/x-www-form-urlencoded",
                           "Cookie" => cookie
                         ),
                         body: form)
        return nil unless resp.ok?

        JSON.parse(clean_json(resp.body))
      rescue StandardError
        nil
      end
    end
  end
end
